test/modules/bradc/userInsteadOfStandard/foo2.chpl

```diff
@@ -1,3 +1,25 @@
+// This test replaces an automatically-included standard library file with
+// a user one because AutoMath.chpl exists as a sibling to this file.
+
+// While this test uses AutoMath, as of this writing, there are 4 automatic
+// standard libraries: AutoMath, Errors, ChapelIO, and Types.
+// A workaround in parser.cpp allows this replacing process for these
+// except for Errors (because it runs into another workaround in the parser).
+// See issue #19313 for a non-workaround proposal to address this.
+
+// One challenge here is that there can only be one copy of a top-level
+// module in a given compilation. That means that if this test uses
+// ./AutoMath.chpl as the AutoMath module, the internal library will
+// refer to it for things like min/max instead of the real one.
+
+// A related challenge to that, and part of the issue with Errors, is that in
+// order for this to work, the first use of AutoMath that the compiler seeks
+// to resolve has to be this one (vs one coming from somewhere else, e.g. an
+// already-parsed included module). That presents an ordering constraint that
+// causes problems when doing incremental or separate compilation. This issue,
+// combined with the limited number of modules to which this can apply, make
+// it potentially not worthwhile to support this pattern.
```

Comment:
I disagree that the pattern is not worthwhile to support.  I have found replacing standard modules incredibly useful to demonstrate and track issues with the standard modules themselves.  It allows us to write a test with the failure behavior without actually exposing the failure behavior to general user programs, breaking other tests, or forgetting about the problem until the next time we encounter it.  Having the issue tracked in a future alerts us when the problem is resolved, allowing us to replace suboptimal patterns that were introduced to avoid said bugs quickly.  Such benefits are worth the additional implementation cost, in my opinion, especially while we are still working to make these modules stable.

The version of the module used is only adjusted by the presence of a compilation flag.  There are other flags that I would expect to make previous incremental compilation results unhelpful (e.g. the addition or removal of `--fast` should have a broader impact than just whichever file happens to also be modified since the last time the program was compiled).  I personally don't see a problem with the addition or removal of this flag causing a more fresh compilation than would otherwise occur, for similar reasons.

I thought it was part of the goal of separate compilation to allow pieces compiled with different versions of a library to be able to interact?  E.g. if module B uses version 1 of module A and module C uses version 2, I thought separately compiling them would silo both versions of A so that it wouldn't be a problem?  It seems reasonable to me for the presence of this flag to result in the creation of a different version of the standard libraries that a program would link to in a similar way.  It's been a while since we discussed the feature, though so I may be misremembering.

Reply:
> The version of the module used is only adjusted by the presence of a compilation flag.

That is not true for this test. Instead, it's just the existence of the AutoMath.chpl module in the same directory, combined with the `use AutoMath` in the test. IMO neither of these inherently cause "Use this instead of the standard library". In particular, if we were to compile another file first (in another directory) in an incremental compilation session, I don't see any reason why we would choose this AutoMath instead of the one in the standard library.

I would find this feature less troubling if it were activated by a compilation flag e.g. `--replace-std-module AutoMath path/to/my/AutoMath.chpl`.

Reply:
Ohhhh, good point.  I agree that it shouldn't trigger without a compilation flag

Reply:
(Another potential compiler flag would be `--prefix-std-module-search-path path/to/my/replacement/` )